src/solver/conv_asm_1x1u.cpp

```diff
@@ -243,75 +244,182 @@ bool PerformanceConfigConvAsm1x1U::operator==(const PerformanceConfigConvAsm1x1U
+    {
+        if(!(k_mult * waves_k_in_group <= problem.n_outputs))
+            return false;
+        if(!(waves_c_in_group * waves_k_in_group <= 16))
+            return false;
+        const auto max_waves_per_CU = (256 / vgprs) * 4;
+        if(!(max_waves_per_CU >= waves_c_in_group * waves_k_in_group))
+            return false;
+    }
+    return true;
+}
 
-    const int total_chunks = (img_hw + chunk_size - 1) / chunk_size;
-    if(!(chunks_per_wave <= total_chunks))
-        return false;
+bool PerformanceConfigConvAsm1x1U::TryToken(int index, int value, const ProblemDescription& problem)
+{
+    switch(index)
+    {
+    case 0: read_size = value; break;
+    case 1: k_mult = value; break;
+    case 2: chunks_per_wave = value; break;
+    case 3: chunk_size = value; break;
+    case 4: n_mult = value; break;
+    case 5: c_mult = value; break;
+    case 6: waves_c_in_group = value; break;
+    case 7: waves_k_in_group = value; break;
+    default: return false;
+    }
+    return this->IsValid(problem);
+}
 
-    const int c_per_wave      = (problem.n_inputs + waves_c_in_group - 1) / waves_c_in_group;
-    const int c_per_last_wave = problem.n_inputs - (c_per_wave * (waves_c_in_group - 1));
+bool IsModelApplicable(const ConvolutionContext& ctx, const ProblemDescription& problem)
+{
+#if MIOPEN_ENABLE_AI_HEUR
+    if(ctx.GetStream().GetDeviceName() == "gfx908" && problem.kernel_stride_h == 1)
+        return true;
+#endif
+    return false;
+}
 
-    if(problem.direction.IsBackwardData() && !(problem.n_outputs % k_mult == 0))
-        return false;
-    return (c_per_wave % c_mult == 0) && (c_per_last_wave % c_mult == 0);
+std::vector<float> TransformFeatures(const ProblemDescription& problem, const int& n)
+{
+    std::vector<float> features(static_cast<unsigned long>(n * n), 0.0);
+    features[0]                   = problem.IsFp32() ? 2.0 : 1.0;
+    int offset                    = (problem.direction.IsForward() ? 0 : 1) + 1;
+    features[(offset)*n + offset] = 1.0;
+    features[3 * n + 3] =
+        float(problem.direction.IsForward() ? problem.n_inputs : problem.n_outputs);
+    features[4 * n + 4] =
+        float(problem.direction.IsForward() ? problem.n_outputs : problem.n_inputs);
+    features[5 * n + 5] = float(problem.in_height);
+    features[6 * n + 6] = float(problem.in_width);
+    features[7 * n + 7] = float(problem.batch_sz);
+    return features;
 }
 
-void PerformanceConfigConvAsm1x1U::HeuristicInit(const ProblemDescription& problem)
+void PerformanceConfigConvAsm1x1U::HeuristicInit(const ConvolutionContext& ctx,
+                                                 const ProblemDescription& problem)
 {
     if(problem.in_data_type == miopenDouble)
         MIOPEN_THROW("Double data type is not supported by ConvAsm1x1U");
 
+    if(IsModelApplicable(ctx, problem))
+    {
+        std::string base_file_path =
+            GetSystemDbPath() + "/" + ctx.GetStream().GetDeviceName() + "_ConvAsm1x1U_";
```

Comment:
This detail should not get into the solvers' implementation. We need to design an interface between solvers and AI tuning. It can be done later, but until that I highly recommend assigning this feature an experimental status and leaving it OFF by default.

Reply:
I wrote a few api functions that abstracted this detail away from the solvers. Let me know if it is enough.